src/config.rs

```diff
@@ -182,19 +182,19 @@ pub struct StyleConfig {
     pub example_variable: Style,
 }
 
-#[derive(Copy, Clone, Debug, PartialEq)]
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
 pub struct DisplayConfig {
     pub compact: bool,
     pub use_pager: bool,
 }
 
-#[derive(Copy, Clone, Debug, PartialEq)]
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
 pub struct UpdatesConfig {
     pub auto_update: bool,
     pub auto_update_interval: Duration,
 }
 
-#[derive(Copy, Clone, Debug, PartialEq)]
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
```

Comment:
Yeah, I guess I see why you did this, but I am not sure it's the right approach. Implementing `Default` for all the configs makes sense, but only if it is a hand written implementation where you would have to manually think about what a default should be (for example when you add a new flag, you don't want to forget that you have to provide a default).

I do like that this test is there though. So what do we do :thinking: ? Firstly, I noticed that `highlight_command` doesn't need the whole config at all, so taking just a `StyleConfig` is an easy refactor (same goes for `format_code` and maybe even further up the stack, but I didn't check). That would still make it rather difficult to test. I would go even further and separate the highlighting logic from the config logic completely. I am thinking

```rust
fn highlight_command<'a, NamePainter, CodePainter>(
    command: &'a str,
    example_code: &'a str,
    // config: &Config,
    paint_name: NamePainter,
    paint_code: CodePainter,
    parts: &mut Vec<ANSIString<'a>>,
) where
    NamePainter: FnMut(/* ... */) -> ANSIString<'a>,
    CodePainter: FnMut(/* ... */) -> ANSIString<'a>,
```

And than we use these two functions like `parts.push(paint_name(command))`. To glue this back together with the rest of the program, there is a wrapper (let's say we modify `format_code` similarly and only expose `format_code` as an API to use with a config):

```rust
fn format_code_with_style(command: &str, text: &str, style: &StyleConfig) -> String {
    format_code(
        command,
        text,
        |x| style.command_name.paint(x),
        |x| style.example_code.paint(x),
        |x| style.example_variable.paint(x),
    )
}
```

(obviously with better names than `x`)

This might be a bit of work right now, but I think it would really help decompose `tealdeer` into separate independent modules and it would also allow for easier tests. You could provide closures to new `highlight_command` that count how often they have been called and check that in the tests.

Although this is now deeper into tealdeer and more complicated than what was first thought, but I believe this would make a great improvement! If things don't work out or you want to discuss, feel free to comment :smiley:

Reply:
I agree, extending the business logic (by introducing defaults) only to have a simpler test setup is actually not what one should do. Your proposal is good, so I would like to refactor the code accordingly. 👍

Reply:
@SimplyDanny Since this is a more involved refactor, I would also wait for what @dbrgn has to say.

@dbrgn, what do you think? :smile:

Reply:
A simpler alternative (for now) would be to derive `Default` only for test builds, and then refactor in a separate PR.

Making the `highlight_command` function independent from the config using painter functions does sound like an interesting idea though.

Reply:
(see pr linked above)